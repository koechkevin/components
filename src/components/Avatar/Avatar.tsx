import { Avatar as AntdAvatar } from 'antd';
import React, { FC } from 'react';

import styles from './Avatar.module.scss';

import { Props } from './interface.d';

export const Avatar: FC<Props> = (props) => {
  const { children, className, icon, src, shape, size, style, color } = props;
  const avatarSizes = {
    small: 32,
    medium: 40,
    large: 80,
    default: 'default',
  };
  const avatarSize = size && (avatarSizes[size] || size);
  const placeholderIcon = src ? icon : 'user';
  const classNames: string = [styles.avatar, className ? className : ''].join(' ');
  const avatarColor = color && !src ? { backgroundColor: color } : {};

  return (
    <AntdAvatar
      src={src}
      shape={shape}
      size={avatarSize}
      className={classNames}
      style={{ minWidth: avatarSize, ...avatarColor, ...style }}
      icon={!children ? placeholderIcon : ''}
    >
      {children}
    </AntdAvatar>
  );
};

Avatar.defaultProps = {
  size: 'default',
};

export default Avatar;
