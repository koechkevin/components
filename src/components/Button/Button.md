# Button

/src/components/Button/Button.tsx

``` TS
<Button>This is a Button</Button>
```
