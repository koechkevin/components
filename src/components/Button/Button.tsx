import { faSpinnerThird } from '@fortawesome/pro-duotone-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Button as AntdButton } from 'antd';
import { ButtonProps } from 'antd/lib/button';
import React, { FC } from 'react';

import styles from './Button.module.scss';

const Button: FC<ButtonProps> = (props) => {
  const { loading, children, className, ...resetProps } = props;
  const classNames = [styles.button, loading ? styles.loading : '', className].join(' ');

  return (
    <AntdButton className={classNames} {...resetProps}>
      {loading ? (
        <FontAwesomeIcon style={{ fontSize: 22, animationDuration: '1.2s' }} icon={faSpinnerThird} spin />
      ) : (
        children
      )}
    </AntdButton>
  );
};

export default Button;
