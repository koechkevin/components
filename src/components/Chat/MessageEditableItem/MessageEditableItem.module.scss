@import '../../../styles/global';

.editableItem {
  &:global(.ant-list-item) {
    display: flex;
    padding: 8px 32px;
    align-items: stretch;
    flex-direction: column;
    background: $color-gray-light;
  }

  @include respond-to('sm') {
    padding: 40px 8px 8px 8px !important; // force rewrite .ant-list-item
  }

  :global {
    .ant-list-item-meta-title {
      margin-bottom: 16px;

      @include respond-to('xs') {
        margin-bottom: 10px;
      }
    }

    .ant-list-item-meta-avatar {
      @include respond-to('xs') {
        display: none;
      }
    }

    .ant-list-item-meta-description {
      @include respond-to('xs') {
        display: flex;
      }
    }
  }
}

.timestamp {
  margin-left: 8px;

  &:global(.ant-typography) {
    color: rgba($color: $color-base-dark, $alpha: 0.6);
  }
}

.buttonGroup {
  margin-top: 16px;
  padding-left: 56px;

  button:first-child {
    margin-right: 15px;
  }

  @include respond-to('xs') {
    padding: 0;
  }
}
