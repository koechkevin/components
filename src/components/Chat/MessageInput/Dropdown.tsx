import { faGoogleDrive } from '@fortawesome/free-brands-svg-icons';
import { faLaptop } from '@fortawesome/pro-solid-svg-icons';
import { Dropdown as AntdDropdown, Menu as AntdMenu, notification, Upload } from 'antd';
import { RcFile } from 'antd/lib/upload';
import React, { FC } from 'react';

import { Icon } from '../../Icon';
import { IconMenuItem, Menu } from '../../Menu';
import { Text } from '../../Typography';
import styles from './MessageInput.module.scss';

const { Item } = AntdMenu;

interface Props {
  uploadToGoogle?: () => void;
  onUpload?: (file: RcFile) => void;
}

const Dropdown: FC<Props> = (props) => {
  const { onUpload, uploadToGoogle, children } = props;

  const beforeUpload = (file: RcFile, fileList: RcFile[]): boolean => {
    if (file.size >= 100 * 1024 * 1024) {
      notification.warning({
        message: 'Oops! That file is too large, we only allow files up to 100mb to be uploaded here.',
      });
      return false;
    }

    onUpload && onUpload(file);
    return false;
  };

  return (
    <AntdDropdown
      trigger={['click']}
      overlay={
        <Menu>
          <Item onClick={uploadToGoogle}>
            <IconMenuItem icon={<Icon icon={faGoogleDrive} />}>Google Drive</IconMenuItem>
          </Item>
          <Item style={{ padding: 0 }}>
            <Upload showUploadList={false} multiple={false} className={styles.upload} beforeUpload={beforeUpload}>
              <Icon icon={faLaptop} style={{ marginRight: 16 }} />
              <Text style={{ fontSize: 16 }}>Your Computer</Text>
            </Upload>
          </Item>
        </Menu>
      }
    >
      {children}
    </AntdDropdown>
  );
};

export default Dropdown;
