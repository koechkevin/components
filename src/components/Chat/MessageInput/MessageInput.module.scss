@import '../../../styles/global';

.messageInput {
  width: 100%;
  position: relative;
  background-color: $color-white;
  min-height: 48px;

  &:focus-within,
  &:hover {
    .mentions {
      border-color: $color-primary;
    }
  }
}

.mentions {
  flex-wrap: nowrap;
  border-radius: 6px;
  align-items: stretch;
  border: 1px solid $color-gray-regular;
  min-height: 48px;

  :global(.ant-col) {
    display: flex;
    align-items: center;
  }

  .message {
    flex: 1;
    width: 100%;
    padding-left: 10px;
    border-width: 0;
    outline: 0;
    resize: none;
    color: $color-base-dark;
    vertical-align: middle;

    @include respond-to('xs-l') {
      font-size: 16px;
    }
  }
}

.addButton {
  margin: 7px;
  width: 32px;
  height: 32px;
  line-height: 1;
  font-size: 16px;
  border-radius: 4px;
  cursor: pointer;
  transition: background 0.2s ease-in;
  @include flex-center;

  &:hover {
    color: $color-primary;
    background-color: $color-primary-light;
  }
}

.iconButton {
  display: flex;
  cursor: pointer;
  box-sizing: border-box;
  color: $color-base-gray-dark;

  > svg {
    margin-left: 10px;
    font-size: 16px;
  }

  &:hover {
    color: $color-primary;
  }
}

.tipsWrapper {
  font-size: 11px;
  text-align: right;

  .tipsContent {
    padding: 4px 0;
    margin-bottom: 0;

    @include respond-to('md') {
      display: none;
    }
  }

  .tipsIcon {
    padding: 4px 0;
    display: none;
    cursor: pointer;
    color: $color-primary;

    @include respond-to('md') {
      display: inline-block;
    }
  }
}

.messageTips {
  margin-bottom: 0;

  @include ellipsis;

  .tisContent span {
    color: $color-primary;
  }
}

.tooltip {
  max-width: 420px;
  font-size: 11px;

  :global(.ant-tooltip-inner) {
    padding: 15px 12px;
    color: $color-base-dark-gray;
    background-color: $color-white;
  }

  :global(.ant-tooltip-arrow) {
    border-top-color: $color-white;
  }
}

.suggestions {
  position: absolute;
  padding-top: 12px;
  bottom: 55px;
  z-index: 999;
  margin-left: -30px;
  margin-bottom: -9px;
  border-radius: 4px;
  background: $color-white;
  box-shadow: 0 4px 12px 0 rgba(0, 0, 0, 0.08), 0 0 0 1px rgba(0, 0, 0, 0.08);

  @include respond-to('sm') {
    bottom: 63px;
  }

  .matches {
    padding: 4px 9px;
    font-size: 10px;
    background: #f5f5f5;
    color: $color-base-dark-gray;

    b {
      margin-left: 4px;
      color: $color-base-dark;
    }
  }
}

.upload {
  display: flex;

  :global {
    .ant-upload {
      width: 100%;
      height: 100%;
      color: $color-base-dark;

      .ant-upload {
        display: flex;
        align-items: center;
        padding: 8px 16px;
      }
    }
  }
}
