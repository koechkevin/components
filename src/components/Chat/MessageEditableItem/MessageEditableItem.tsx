import { List, Row } from 'antd';
import React, { FC, useState } from 'react';

import { MessageItemProps, SuggestionsUser } from '../../../typings/Message';
import { getNameInitials, ProductType, shortTimeStampFormat } from '../../../utils';
import { Avatar } from '../../Avatar';
import { Button } from '../../Button';
import { Text } from '../../Typography';
import MessageInput from '../MessageInput/MessageInput';

import styles from './MessageEditableItem.module.scss';

const { Item } = List;

export interface EditableMessageProps extends MessageItemProps {
  suggestions: SuggestionsUser[];
  save: (text: string) => void;
  cancel: (messageId: string) => void;
}

const EditableItem: FC<EditableMessageProps> = (props) => {
  const { id, author, text, cancel, save, suggestions, createdAt } = props;
  const [msg, setMsg] = useState(text);

  return (
    <Item className={styles.editableItem}>
      <Item.Meta
        avatar={
          <Avatar
            size={40}
            src={author && author.avatar}
            status={author && author.chatStatus}
            shape={author && author.productId === ProductType.Candidate ? 'square' : 'circle'}
            style={{ background: author && author.avatarColor, fontWeight: 'bold' }}
          >
            {author && getNameInitials(author.name)}
          </Avatar>
        }
        title={
          <Row>
            <Text strong>{author ? author.name : ''}</Text>
            <Text className={styles.timestamp}>{createdAt && shortTimeStampFormat(new Date(createdAt))}</Text>
          </Row>
        }
        description={
          <MessageInput
            text={msg}
            showAt={false}
            onEnter={save}
            suggestions={suggestions}
            onChange={(text: string) => setMsg(text)}
          />
        }
      />
      <Row className={styles.buttonGroup}>
        <Button ghost onClick={() => id && cancel(id)}>
          Cancel
        </Button>
        <Button type="primary" onClick={() => msg && save(msg)}>
          Save Changes
        </Button>
      </Row>
    </Item>
  );
};

export default EditableItem;
