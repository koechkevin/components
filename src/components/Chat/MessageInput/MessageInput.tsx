import { faAt, faInfoCircle, faSmile } from '@fortawesome/pro-light-svg-icons';
import { faPlus } from '@fortawesome/pro-regular-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { Col, Row } from 'antd';
import { RcFile } from 'antd/lib/upload';
import React, { ChangeEvent, FC, KeyboardEvent, MouseEvent, useCallback, useEffect, useRef, useState } from 'react';
import InputTrigger from 'react-input-trigger';
import TextArea from 'react-textarea-autosize';
import { useClickAway, useMedia } from 'react-use';
import { SuggestionsUser } from '../../../typings/Message';

import { Suggestion } from '../../Suggestion';
import { Tooltip } from '../../Tooltip';
import Typing from '../Typing/Typing';
import Dropdown from './Dropdown';

import styles from './MessageInput.module.scss';

const TipsContent: FC<any> = (props: any) => {
  const { ...restProps } = props;
  return (
    <p className={styles.messageTips} {...restProps}>
      Use <span>@</span> for mentions, <span>/</span> for more options , <span>#</span> for channels | <span>+</span>{' '}
      Add files to share
    </p>
  );
};

export interface Props {
  text?: string;
  showAt?: boolean;
  hintShown?: boolean;
  suggestions: SuggestionsUser[];
  typingUsers?: string[];
  onEnter: (text: string) => void;
  onChange?: (text: string) => void;
  onUpload?: (file: RcFile) => void;
}

export const MessageInput: FC<Props> = (props) => {
  const { showAt, hintShown, suggestions, onEnter, onChange, onUpload, typingUsers } = props;
  const content = props.text ? props.text.replace(/<br\/>/g, '\n') : '';
  const [text, setText] = useState(content);
  const isMobile = useMedia('(max-width: 575px)');

  /** element ref */
  const inputRef = useRef<HTMLTextAreaElement | null>(null);
  const suggestionsRef = useRef<HTMLDivElement | null>(null);

  /** mention suggestions */
  const [suggestionLeftPos, setSuggestionLeftPos] = useState<number>(0);
  const [suggestionShown, setSuggestionShown] = useState<boolean>(false);
  const [currentSelection, setCurrentSelection] = useState<number>(0);
  const [endHandler, setEndHandler] = useState<() => void>(() => (): any => null);
  const [keyword, setKeyword] = useState<string>('');
  const [startPos, setStartPos] = useState<number>(0);

  useClickAway(suggestionsRef, () => {
    setSuggestionShown(false);
  });

  const handleChange = useCallback(
    (text: string) => {
      setText(text);

      // replace all `\n` `\r` with `<br />`
      const content = text.replace(/\r?\n/g, '<br/>');

      if (onChange) {
        onChange(content);
      }
    },
    [onChange],
  );

  const handleAtClick = () => {
    handleChange(text + '@');
    setSuggestionShown(true);
  };

  /** send message */
  const onReply = () => {
    if (text.trim() !== '') {
      onEnter(text);
      setText('');
    }
  };

  /** handle textarea keydown */
  const onKeyDown = (evt: KeyboardEvent<HTMLTextAreaElement>) => {
    if (evt.key === 'Enter') {
      if (suggestionShown) {
        return;
      }

      if (evt.ctrlKey) {
        if (inputRef && inputRef.current) {
          const pos = inputRef.current?.selectionEnd;
          const newText = text.substring(0, pos) + '\n' + text.substring(pos);
          setText(newText);
        }
      }

      if (evt.shiftKey || evt.ctrlKey) {
        evt.stopPropagation();
      } else {
        evt.preventDefault();
        onReply();
      }
    }
  };

  const reset = useCallback(() => {
    setCurrentSelection(0);
    setSuggestionLeftPos(0);
    setSuggestionShown(false);
    endHandler();
    setStartPos(0);
  }, [endHandler]);

  const toggleSuggestion = (meta: any) => {
    const { hookType, cursor } = meta;

    if (hookType === 'start') {
      setSuggestionShown(true);
      setSuggestionLeftPos(cursor.left);
      setStartPos(cursor.selectionStart);
      setKeyword('');
    }

    if (hookType === 'cancel') {
      reset();
    }
  };

  const handleInput = (meta: any) => {
    const { text } = meta;
    setKeyword(text);
  };

  const done = useCallback(() => {
    const users = suggestions.filter((user: SuggestionsUser) => user.name.toLowerCase().indexOf(keyword) !== -1);
    const user = users[currentSelection];

    if (user) {
      const username = `@${user.name}`;
      const newText =
        `${text.slice(0, startPos > 1 ? startPos - 1 : 0)}` +
        `${username}${text.slice(startPos + username.length, text.length)}`;
      handleChange(newText);
    }
    reset();
  }, [currentSelection, handleChange, suggestions, reset, keyword, text, startPos]);

  const handleKeyDown = (evt: React.KeyboardEvent) => {
    const { which, shiftKey } = evt;

    if (suggestionShown && which === 40) {
      evt.preventDefault();
      setCurrentSelection((currentSelection + 1) % suggestions.length);
    }

    if (suggestionShown && which === 38 && currentSelection > 0) {
      evt.preventDefault();
      setCurrentSelection((currentSelection - 1) % suggestions.length);
    }

    if (which === 13) {
      if (!shiftKey) {
        if (suggestionShown) {
          evt.preventDefault();
          done();
        }
      } else {
        reset();
      }
    }
  };

  const handleSuggestionItemClick = (e: MouseEvent, idx: number) => {
    e.stopPropagation();
    setCurrentSelection(idx);
    done();

    if (inputRef && inputRef.current) {
      inputRef.current.focus();
    }
  };

  const handleHover = (e: MouseEvent, idx: number) => {
    e.stopPropagation();
    setCurrentSelection(idx);
  };

  useEffect(() => {
    const range = text.length;

    if (inputRef && inputRef.current) {
      inputRef.current.setSelectionRange(range, range);
    }
  }, [inputRef]);

  const filtered: SuggestionsUser[] = suggestions
    ? suggestions.filter((user: SuggestionsUser) => user.name.toLowerCase().indexOf(keyword) !== -1)
    : [];

  return (
    <Row className={styles.messageInput}>
      <Row type="flex" align="top" className={styles.mentions}>
        <Col style={{ alignItems: 'flex-start', padding: '14px 0' }}>
          {showAt && (
            <span ref={suggestionsRef} className={styles.iconButton} onClick={() => handleAtClick()}>
              <FontAwesomeIcon icon={faAt} style={{ fontSize: 16 }} />
            </span>
          )}
          <span className={styles.iconButton}>
            <FontAwesomeIcon icon={faSmile} />
          </span>
        </Col>
        <Col style={{ flex: 1 }} onKeyDown={handleKeyDown}>
          <InputTrigger
            trigger={{
              keyCode: 50 || 51,
              shiftKey: true,
            }}
            inputRef={() => inputRef}
            onType={(meta: any) => handleInput(meta)}
            style={{ width: '100%', padding: '10px 0' }}
            onStart={(meta: any) => toggleSuggestion(meta)}
            onCancel={(meta: any) => toggleSuggestion(meta)}
            endTrigger={(endHandler: any) => setEndHandler(() => endHandler)}
          >
            <TextArea
              minRows={1}
              value={text}
              inputRef={inputRef}
              onKeyDown={onKeyDown}
              maxRows={isMobile ? 5 : 8}
              className={styles.message}
              autoFocus={isMobile ? false : true}
              onChange={(e: ChangeEvent<HTMLTextAreaElement>) => handleChange(e.target.value)}
            />
          </InputTrigger>
          {/** suggestion box */}
          {suggestionShown && (
            <div
              className={styles.suggestions}
              ref={suggestionsRef}
              style={{ left: isMobile ? '' : suggestionLeftPos }}
            >
              {filtered.map((user: SuggestionsUser, idx: number) => (
                <Suggestion
                  key={idx}
                  inCall={Boolean(user.inCall)}
                  isActive={idx === currentSelection}
                  signature={user.signature}
                  onMouseOver={(e: MouseEvent) => handleHover(e, idx)}
                  onClick={(e: MouseEvent) => handleSuggestionItemClick(e, idx)}
                  {...user}
                />
              ))}
            </div>
          )}
        </Col>
        <Col style={{ alignItems: 'flex-start' }}>
          <Dropdown onUpload={onUpload}>
            <span className={styles.addButton}>
              <FontAwesomeIcon icon={faPlus} />
            </span>
          </Dropdown>
        </Col>
      </Row>

      <div className={styles.tipsWrapper}>
        {!isMobile && (
          <Typing
            visibility={typingUsers && typingUsers.length !== 0 ? 'visible' : 'hidden'}
            typingUsers={typingUsers}
          />
        )}
        {hintShown && (
          <>
            <TipsContent className={styles.tipsContent} />
            <Tooltip arrowPointAtCenter title={<TipsContent />} overlayClassName={styles.tooltip}>
              <span className={styles.tipsIcon}>
                <FontAwesomeIcon icon={faInfoCircle} />
                Tips
              </span>
            </Tooltip>
          </>
        )}
      </div>
    </Row>
  );
};

MessageInput.defaultProps = {
  showAt: true,
};

export default MessageInput;
