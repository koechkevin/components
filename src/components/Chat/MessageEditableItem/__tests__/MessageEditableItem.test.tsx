import { mount } from 'enzyme';
import React from 'react';

import { Avatar } from '../../../Avatar/Avatar';
import Button from '../../../Button/Button';
import MessageInput from '../../MessageInput/MessageInput';
import MessageEditableItem from '../MessageEditableItem';

describe('Tests for message editable item component', () => {
  const props = {
    id: 'sample id',
    threadId: 'sampleThreadId',
    createdAt: '2019-10-08 06:22:09',
    author: {
      id: 'user id',
      avatar: 'sample avatar',
      status: 'sample status',
      isRobot: false,
      avatarColor: '#4224',
      name: 'Jane Doe',
      productId: 'company-portal',
    },
    text: 'sample text',
    suggestions: [
      {
        id: 'suggestion id',
        name: 'sample name',
        avatar: 'sample avatar',
        avatarColor: '#FFF',
        status: 'sample status',
        isRobot: false,
      },
    ],
    save: jest.fn(),
    cancel: jest.fn(),
  };

  const wrapper = mount(<MessageEditableItem {...props} />);

  it('should test for the mounting of the message editable component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should be able to render circular avatar for company profile', () => {
    const avatar = wrapper.find(Avatar);

    expect(avatar.props().shape).toEqual('circle');
  });

  it('should render square avatar for candidate profile', () => {
    const candidateProps = {
      ...props,
      author: {
        ...props.author,
        productId: 'candidate-portal',
      },
    };
    const candidateWrapper = mount(<MessageEditableItem {...candidateProps} />);
    const avatar = candidateWrapper.find(Avatar);

    expect(avatar.props().shape).toEqual('square');
  });

  it('should simulate onChange for message input', () => {
    const messageInput = wrapper.find(MessageInput);

    messageInput.simulate('change');

    expect(messageInput.props().text).toEqual('sample text');
  });

  it('should render author name', () => {
    expect(wrapper.contains(props.author.name)).toEqual(true);
  });

  it('should save message on click of save button', () => {
    const button = wrapper.find(Button).last();

    expect(button.props().type).toEqual('primary');
    expect(button.contains('Save Changes')).toEqual(true);

    button.simulate('click');

    expect(props.save).toHaveBeenCalled();
  });

  it('should simulate cancelling on message editable component', () => {
    const cancelButton = wrapper.find(Button).first();

    expect(cancelButton.props().ghost).toEqual(true);
    expect(cancelButton.contains('Cancel')).toEqual(true);

    cancelButton.simulate('click');

    expect(props.cancel).toHaveBeenCalled();
  });
});
