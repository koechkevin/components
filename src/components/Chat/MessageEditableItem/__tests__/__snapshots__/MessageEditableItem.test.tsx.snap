// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for message editable item component should test for the mounting of the message editable component 1`] = `ReactWrapper {}`;
